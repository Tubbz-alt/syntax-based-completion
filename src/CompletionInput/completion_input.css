.CompletionInput,
.CompletionInput * {
    box-sizing: border-box;
}

.CompletionInput {
    position: relative;
}

.CompletionInput__input--error {
    border: 1px solid #d0203d;
    outline-color: #d0203d;
}

.CompletionInput__error {
    position: absolute;
    background: #d0203d;
    width: 100%;
    display: block;
    color: #fff;
    font-size: .8rem;
    padding: .2rem .5rem;
    margin-top: -1px;
}
