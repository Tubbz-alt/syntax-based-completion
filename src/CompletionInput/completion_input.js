import React from "react";
import PropTypes from "prop-types";
import { Parser } from "../utils";

import "./completion_input.css";

import SuggestionList from "../SuggestionList";
import { COMMANDS, DELAY, STEPS } from "../constants";

export default class CompletionInput extends React.Component {
  static propTypes = {
    syntax: PropTypes.string.isRequired,
    value: PropTypes.string,
    delay: PropTypes.number,
    onChange: PropTypes.func,
    fetchSuggestions: PropTypes.func.isRequired,
    className: PropTypes.className,
    readOnly: PropTypes.boolean,
    placeholder: PropTypes.string
  };

  static defaultProps = {
    className: "",
    delay: DELAY,
    value: ""
  };

  fetcher;
  parser;
  input;

  constructor(props) {
    super(props);

    this.input = null;
    this.fetcher = null;

    this.state = {
      suggestions: [],
      fetchingSuggestions: false,
      cursorPosition: props.value.length,
      focusingList: false,
      active: 0,
      showSuggestions: false,
      error: null,
      output: null
    };
  }

  componentDidMount() {
    this.setupParser();
  }

  // Helpers

  setupParser = async () => {
    this.parser = new Parser(this.props.syntax);
    const output = this.getOutput(this.props.value);
    const error = this.checkForErrors(output);

    this.setState({
      output,
      error
    });
  };

  checkForErrors = output => {
    if (output.errors && output.errors.length) {
      return output.errors[0].message;
    }
    return null;
  };

  delayAction = async action => {
    clearTimeout(this.fetcher);
    this.fetcher = setTimeout(action, this.props.delay);
  };

  getOutput = sentence => {
    const shouldUpdateOutput =
      !this.state.output || sentence !== this.state.output.text;
    const output = shouldUpdateOutput
      ? this.parser.parse(sentence)
      : this.state.output;

    return output;
  };

  updateReferences = async (silent = false) => {
    const sentence = this.input.value;
    const cursorPosition = isFinite(this.input.selectionStart)
      ? this.input.selectionStart
      : sentence.length;
    const changedCursor = cursorPosition !== this.state.cursorPosition;
    const shouldUpdateOutput =
      !this.state.output || sentence !== this.state.output.text;

    if (!silent) {
      this.setState({
        loading: true
      });
    }

    if (changedCursor || shouldUpdateOutput) {
      this.delayAction(() => {
        const output = this.getOutput(sentence);
        const error = this.checkForErrors(output);

        this.delayAction(() => {
          const currentNode = error
            ? output
            : output.findChildByPosition(cursorPosition);

          this.setState({
            cursorPosition,
            output,
            error,
            currentNode,
            showSuggestions:
              silent || error ? false : this.state.showSuggestions,
            active: 0
          });

          if (!silent && !error) this.checkForSuggestions(currentNode);
        });
      });
    }
  };

  sortSuggestions = suggestions => {
    const { text } = this.state.currentNode;

    if (!suggestions || suggestions.length === 0) return [];
    if (!text) return suggestions;

    var results = suggestions.reduce(
      function(res, suggestion) {
        var pos = suggestion.toLowerCase().indexOf(text.toLowerCase());
        if (pos === 0) {
          res[0].push(suggestion);
        } else if (pos > 0) {
          res[1].push(suggestion);
        } else {
          res[2].push(suggestion);
        }
        return res;
      },
      [[], [], []]
    );
    return [...results[0], ...results[1], ...results[2]];
  };

  checkForSuggestions = async (currentNode = this.state.currentNode) => {
    this.setState({
      showSuggestions: true,
      fetchingSuggestions: true
    });

    if (!!this.props.fetchSuggestions) {
      this.delayAction(() => {
        this.props
          .fetchSuggestions(currentNode)
          .then((suggestions = []) => {
            this.setState({
              suggestions: this.sortSuggestions(suggestions),
              fetchingSuggestions: false
            });
          })
          .catch(() => {
            this.setState({
              suggestions: [],
              showSuggestions: false,
              fetchingSuggestions: false
            });
          });
      });
    }
  };

  navigateDown = () => {
    const { active } = this.state;
    const lastPosition = this.state.suggestions.length - 1;
    return active === lastPosition ? active : active + STEPS;
  };

  navigateUp = () => {
    const { active } = this.state;
    return active === 0 ? active : active - STEPS;
  };

  // Handlers

  onMouseEnterList = () => {
    this.setState({
      focusingList: true
    });
  };

  onMouseLeaveList = () => {
    this.setState({
      focusingList: false
    });
  };

  onClose = () => {
    clearTimeout(this.fetcher);

    this.setState({
      fetchingSuggestions: false,
      showSuggestions: false,
      suggestions: [],
      active: 0
    });
  };

  onPressEnter = () => {
    const selected = this.state.suggestions[this.state.active];
    if (typeof selected !== "undefined") this.onSelect(selected);
  };

  onNavigate = e => {
    const active =
      e.key === COMMANDS.DOWN ? this.navigateDown() : this.navigateUp();

    this.setState({ active });
    e.stopPropagation();
    e.preventDefault();
  };

  onChange = () => {
    this.delayAction(this.updateReferences);
    if (this.props.onChange) this.props.onChange(this.input.value);
  };

  onKeyUp = e => {
    switch (e.key) {
      case COMMANDS.LEFT:
      case COMMANDS.RIGHT:
        this.updateReferences();
        break;
      case COMMANDS.UP:
      case COMMANDS.DOWN:
        this.onNavigate(e);
        break;
      case COMMANDS.ENTER:
        this.onPressEnter();
        break;
      case COMMANDS.ESC:
        this.onClose();
        break;
      default:
        break;
    }
  };

  onMouseUp = e => {
    if (e.target === this.input) this.updateReferences();
  };

  onBlur = () => {
    if (!this.state.focusingList) this.onClose();
  };

  onSelect = (newValue = "") => {
    const sentence = this.props.value;
    const node = this.state.currentNode;
    const newSentence =
      "" +
      sentence.substring(0, node.start) +
      newValue +
      sentence.substring(node.end, sentence.length);

    const newSelectionStart = node.start + String(newValue).length;

    this.input.value = newSentence;

    this.input.focus();
    this.input.setSelectionRange(newSelectionStart, newSelectionStart);
    this.onClose();

    if (this.props.onChange) this.props.onChange(newSentence);
  };

  // Renders

  renderSuggestions = () => {
    const { suggestions, showSuggestions } = this.state;

    if (!showSuggestions) {
      return;
    }

    return (
      <SuggestionList
        suggestions={suggestions}
        loading={this.state.fetchingSuggestions}
        active={this.state.active}
        selected={this.state.currentNode ? this.state.currentNode.text : null}
        onSelect={this.onSelect}
        onMouseEnter={this.onMouseEnterList}
        onMouseLeave={this.onMouseLeaveList}
      />
    );
  };

  renderErrors = () => {
    if (this.state.error) {
      return <div className="CompletionInput__error">{this.state.error}</div>;
    }
    return;
  };

  render() {
    const {
      onChange, // eslint-disable-line no-unused-vars
      syntax, // eslint-disable-line no-unused-vars
      fetchSuggestions, // eslint-disable-line no-unused-vars
      ...props
    } = this.props;

    return (
      <div className="CompletionInput">
        <input
          {...props}
          className={
            this.props.className +
            " CompletionInput__input" +
            (this.state.error ? "--error" : "")
          }
          ref={node => {
            this.input = node;
          }}
          readOnly={!this.parser || this.props.readOnly}
          placeholder={
            !this.parser ? "Parsing syntax..." : this.props.placeholder
          }
          onBlur={this.onBlur}
          onMouseUp={this.onMouseUp}
          onKeyUp={this.onKeyUp}
          onChange={this.onChange}
        />
        {this.renderErrors()}
        {this.renderSuggestions()}
      </div>
    );
  }
}
