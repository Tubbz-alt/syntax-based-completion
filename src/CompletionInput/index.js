import CompletionInput from "./completion_input";
export default CompletionInput;
