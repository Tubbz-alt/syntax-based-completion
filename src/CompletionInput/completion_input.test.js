import React from "react";
import { shallow, mount } from "enzyme";

import CompletionInput from "./completion_input";

const SYNTAX = `
<SYNTAX>           ::= <composition>
<composition>      ::= <word> <BREAK> <composition> | <word> <BREAK> | <word>
<word>             ::= <CHARS> | <tag>
<tag>              ::= "#" <CHARS> | "#"
<CHARS>            ::= <CHAR> <CHARS> | <CHAR>
<CHAR>             ::= <LETTER> | <DIGIT> | "_" | "-"
<LETTER>           ::= <UPPER_LETTER> | <LOWER_LETTER>
<UPPER_LETTER>     ::= "A" | "B" | "C" | "D" | "E" | "F" | "G" | "H" | "I" | "J" | "K" | "L" | "M" | "N" | "O" | "P" | "Q" | "R" | "S" | "T" | "U" | "V" | "W" | "X" | "Y" | "Z"
<LOWER_LETTER>     ::="a" | "b" | "c" | "d" | "e" | "f" | "g" | "h" | "i" | "j" | "k" | "l" | "m" | "n" | "o" | "p" | "q" | "r" | "s" | "t" | "u" | "v" | "w" | "x" | "y" | "z"
<DIGIT>            ::= "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
<BREAK>            ::= <BREAK_SYMBOL> <BREAK> | <BREAK_SYMBOL>
<BREAK_SYMBOL>     ::= "," | "." | ";" | " "
`;

const defaultProps = {
  syntax: SYNTAX,
  delay: 0
};

describe("<CompletionInput>", () => {
  describe("shallow render", () => {
    it("parsing syntax", () => {
      const wrapper = shallow(<CompletionInput {...defaultProps} />);
      wrapper.instance().parser = null;
      wrapper.update();
      expect(
        wrapper.find(".CompletionInput__input").prop("placeholder")
      ).toEqual("Parsing syntax...");
      expect(wrapper).toMatchSnapshot();
    });

    it("empty", () => {
      const wrapper = shallow(<CompletionInput {...defaultProps} />);

      expect(wrapper).toMatchSnapshot();
    });

    it("with text", () => {
      const wrapper = shallow(
        <CompletionInput {...defaultProps} value="Foo #bar" />
      );
      expect(wrapper).toMatchSnapshot();
    });

    it("with error", () => {
      const wrapper = shallow(
        <CompletionInput {...defaultProps} value="Foo /Invalid" />
      );
      expect(wrapper.find(".CompletionInput__error")).toHaveLength(1);
      expect(wrapper).toMatchSnapshot();
    });

    it("showing suggestions", () => {
      const wrapper = shallow(<CompletionInput {...defaultProps} />);

      wrapper.setState({ showSuggestions: true, suggestions: ["Foo", "Bar"] });
      wrapper.update();
      expect(wrapper.find("SuggestionList")).toHaveLength(1);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("helpers", () => {
    describe("setupParser", () => {
      const mockSetupParser = jest.fn();
      class MockedInput extends CompletionInput {
        setupParser = mockSetupParser;
      }

      it("is called on mount", () => {
        mount(<MockedInput {...defaultProps} />);
        expect(mockSetupParser).toHaveBeenCalled();
      });
    });

    describe("delayAction", () => {
      const oldClearTimeout = global.clearTimeout;
      const oldSetTimeout = global.setTimeout;

      beforeEach(() => {
        global.clearTimeout = jest.fn();
        global.setTimeout = jest.fn();
      });

      afterEach(() => {
        global.setTimeout = oldSetTimeout;
        global.clearTimeout = oldClearTimeout;
      });

      it("trigger clearTimout/setTimeout", () => {
        const wrapper = shallow(<CompletionInput {...defaultProps} />);
        wrapper.instance().delayAction(() => {});
        expect(global.clearTimeout).toHaveBeenCalled();
        expect(global.setTimeout).toHaveBeenCalled();
      });

      it("updates the same reference", () => {
        const wrapper = shallow(<CompletionInput {...defaultProps} />);
        global.setTimeout.mockImplementation(() => "mock-reference");
        wrapper.instance().fetcher = "initial-mock-reference";

        wrapper.instance().delayAction("action");

        expect(wrapper.instance().fetcher).toBe("mock-reference");
        expect(global.clearTimeout).toHaveBeenCalledWith(
          "initial-mock-reference"
        );
      });
    });
  });

  describe("handlers", () => {
    it("onSelect", () => {});
  });
});
