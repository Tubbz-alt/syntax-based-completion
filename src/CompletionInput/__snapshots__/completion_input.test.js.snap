// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CompletionInput> shallow render empty 1`] = `
<div
  className="CompletionInput"
>
  <input
    className=" CompletionInput__input"
    delay={0}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    onMouseUp={[Function]}
    placeholder="Parsing syntax..."
    readOnly={true}
    value=""
  />
</div>
`;

exports[`<CompletionInput> shallow render parsing syntax 1`] = `
<div
  className="CompletionInput"
>
  <input
    className=" CompletionInput__input"
    delay={0}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    onMouseUp={[Function]}
    placeholder="Parsing syntax..."
    readOnly={true}
    value=""
  />
</div>
`;

exports[`<CompletionInput> shallow render showing suggestions 1`] = `
<div
  className="CompletionInput"
>
  <input
    className=" CompletionInput__input"
    delay={0}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    onMouseUp={[Function]}
    value=""
  />
  <SuggestionList
    active={0}
    loading={false}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onSelect={[Function]}
    selected={null}
    suggestions={
      Array [
        "Foo",
        "Bar",
      ]
    }
  />
</div>
`;

exports[`<CompletionInput> shallow render with error 1`] = `
<div
  className="CompletionInput"
>
  <input
    className=" CompletionInput__input--error"
    delay={0}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    onMouseUp={[Function]}
    value="Foo /Invalid"
  />
  <div
    className="CompletionInput__error"
  >
    Unexpected end of input: 
/Invalid
  </div>
</div>
`;

exports[`<CompletionInput> shallow render with text 1`] = `
<div
  className="CompletionInput"
>
  <input
    className=" CompletionInput__input"
    delay={0}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    onMouseUp={[Function]}
    value="Foo #bar"
  />
</div>
`;
